### ISSUE TITLE
Zed downloads NodeJS binary and npm packages from Internet without user’s consent

### ISSUE THREAD

**Issue by jirutka** [zed-industries/zed#12589]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

I noticed that Zed automatically downloads the NodeJS binary from https://nodejs.org without asking or even informing the user about it. Right after starting it and opening a file, without doing anything else. And there’s no option to disable it.

This is completely unacceptable!

Not just for security reasons but also from a usability point of view. I’m currently connected via metered LTE, and Zed has just eaten up 14 MiB of my plan. Moreover, I already have node installed and on PATH. Also, the downloaded binary is somehow corrupted and it wouldn’t work on my system anyway because it’s built against glibc (that’s how I noticed it in the first place).

And to make matters worse, if it did work, it would start installing arbitrary packages from npmjs.com via npm and running their scripts. This represents a huge attack vector.

This approach is completely unacceptable for anyone who’s concerned about cybersecurity and for virtually all companies, at least in the EU, because of cybersecurity laws, related certifications and audits.

EDIT: Now I found that it downloads ([here](https://github.com/zed-industries/zed/blob/de8ef081436317689333b29324b3a64343745914/crates/supermaven_api/src/supermaven_api.rs#L189)) even some _proprietary_ binary from https://supermaven.com, i.e. unaudited and unauditable code, without any verification (except TLS)! At least this is not downloaded by default… I hope…

EDIT2: Zed also automatically downloads and executes prebuilt language servers for C#, Clojure, Deno, Elixir, Gleam, GLSL, Lua, Terraform, Toml and Zig. It automatically resolves the latest version available on GitHub and downloads it, again, without any verification.

### Environment

N/A

### Somehow related issues

- #12358
- #12354
- #12703

**Comment by asesh** [comment]
I agree with you, it can cause **severe security issues**. It will also download unsigned executable files and run them w/o any consent or permission. That's why I have language server disabled. You can disable this feature though: [https://zed.dev/docs/configuring-zed](https://zed.dev/docs/configuring-zed)

**Comment by vielmetti** [comment]
Some of this is previously discussed in #7054

**Comment by jirutka** [comment]
> That's why I have language server disabled. You can disable this feature though: https://zed.dev/docs/configuring-zed

Language servers are very useful. I don’t see an option to disable the automatic fetching of dependencies from the internet without disabling the language servers completely.

**Comment by jirutka** [comment]
> "We do not have plans to abandon this approach since there's so much code written to support various frontend tools already, that rewriting those in Rust will take an eternity, so not sure what is actionable here, hence closing."

This is a false dichotomy. Nobody is asking to rewrite everything in Rust and bundle it into the application (on the contrary, I’d ask the opposite). Node.js is a normal system dependency that should be installed by the system’s package manager, or manually (if there’s no package manager in place).

**Comment by someone13574** [comment]
Maybe a popup, which is enabled by default, could ask for permission to download `some-binary` from `some-link`? It wouldn't mitigate the fact that its downloading a binary, but it would at least give the user an option to see whats getting downloaded beforehand. Then in settings there could be three options for `download_permission` or something:

- `ask`: The default setting. Asks the user for permission to download a file.
- `always`: The current behavior. Just download without asking.
- `never`: Never download any files. May break extensions and isn't recommended (but available for users who don't have a glibc system, really value security, or any other reason).

**Comment by someone13574** [comment]
Ideally you would be able to turn off auto-download but still be able to use an lsp if it’s already on the users system.

**Comment by WanderLanz** [comment]
Out of curiosity, what's the state of Deno compatibility?
It could potentially be a sandbox layer until the team can decide on a proper sandbox, and could also potentially be embedded last time I checked.

**Comment by vinicius507** [comment]
Yeah, I think Zed the following would be the most sensible approach to this problem:

- Provide the user with the option to configure their language servers binaries and/or detect them from `PATH`.
- Implement the ability for extensions to bundle/download and install native dependencies.

This way official extensions could provide common language servers such as the `json-language-server` so users could choose between providing their language servers vs downloading and installing them via extensions.

**Comment by crabdancing** [comment]
Yeah, this issue really turned me off using Zed until it's fixed. :/

It tried to run them and then hilariously couldn't because I'm on a NixOS system:

```
[2024-06-16T12:55:07-06:00 ERROR project] failed to start language server "json-language-server": failed to execute npm info subcommand:
stdout: ""
stderr: "Could not start dynamically linked executable: /home/ada/.local/share/zed/node/node-v18.15.0-linux-x64/bin/node\nNixOS cannot run dynamically linked executables intended for generic\nlinux environments out of the box. For more information, see:\nhttps://nix.dev/permalink/stub-ld\n"
```

Frankly, not having the ability to configure it to just use what's in my environment (which I have already audited), or specific paths that I give it, is a dealbreaker.

**Comment by imbev** [comment]
I am also concerned about this. Either I disable LSP or I allow the automatic download of arbitrary binaries from GitHub. This is a dealbreaker, and prevents Zed's use in security conscious environments.

**Comment by sgriff96** [comment]
I don't think anyone is asking for the team to re-write every binary in Rust. We just want to be asked to install the language server or not. If you click no, then you don't install all the necessary things for that language.

**Comment by imbev** [comment]
> I don't think anyone is asking for the team to re-write every binary in Rust. We just want to be asked to install the language server or not. If you click no, then you don't install all the necessary things for that language.

The current behavior prompts on a per-language basis, but https://github.com/zed-industries/zed/issues/12589#issuecomment-2148425181 would be a significant improvement.

**Comment by someone13574** [comment]
> I don't think anyone is asking for the team to re-write every binary in Rust. We just want to be asked to install the language server or not. If you click no, then you don't install all the necessary things for that language.

This has been turned down already: https://github.com/zed-industries/zed/pull/12703#issuecomment-2151119560. As said in the comment it is likely that the extension API will be changed if we don't want to manually make sure every extension does the desired behavior (respect a "no downloads" setting, searching path, gracefully handling when downloads aren't allowed, etc.)

**Comment by shinebayar-g** [comment]
Related https://github.com/zed-industries/zed/issues/9224

**Comment by nukeop** [comment]
Oh no, whole 14MB? And what is this NodeJS thing it's downloading? Sounds fishy...

**Comment by crabdancing** [comment]
@nukeop 

For higher security development environments, we tend not to like applications trying to autonomously pull entire massive ecosystems onto disk. It can dramatically increase attack surface if, for instance, a supply chain attack is of concern. So yes, this sets a very bad precedent. Especially with the 'reputation' NPM has garnered, security-wise.

Also, 14MB is kinda significant if you're stuck with one of the really horrible cell phone plans that exist. For instance, I have to stay under a gigabyte, but it wouldn't surprise me if there are people out there stuck with 500MB or less. Using up bandwidth in a context where it isn't obvious without asking the user is an anti-pattern, whether or not it seems insignificant.

The latter issue is of less importance IMO.

**Comment by nukeop** [comment]
I am going to put forward a hypothesis that this is a really really niche use case and 99.99% of users just want to open a JS/TS file and have it work. And the way OP phrased it ("it's COMPLETELY UNACCEPTABLE") is just nasty, entitled, and tries to make it much more of an issue than it has any right to be. Then someone posts it on reddit to stir up an unnecessary flamewar, just because.

**Comment by crabdancing** [comment]
Those 'niche' users may or may not include security conscious developers who often maintain critical applications for major corporations, and if their development environments are compromised, could result in backdooring literally millions of systems in some cases, by injecting malware before the build server even gets a chance to sign it. So yes, it's niche. You know what else is niche? PCB fabrication. Good luck running a functioning modern civilization without it, though.

**Comment by nukeop** [comment]
Then here's how the OP could have looked:

> Hi, I am a security conscious developer who often maintains critical applications for major corporations. Could you please include an option to confirm downloads of the following binaries? [list of binaries] I would like to be able to download them on an opt-in basis.

Instead of acting like they spat on the grave of his ancestors. It's not hard to be a little more diplomatic.

**Comment by crabdancing** [comment]
@nukeop 

:woman_shrugging: Sorry they offended you, I guess? Or the devs? I don't know OP personally, and don't really have a stake in this point.

**Comment by mrnugget** [comment]
Hey! This ticket making the rounds it seems, so wanted to leave a comment in here acknowledging that this is a big deal for some users. We hear you and we plan on doing something about it, but we haven't gotten around to it yet.

Right now, we have three ways of supporting languages and language servers in Zed:

1. Built-in language support (Rust, Go, ...)
2. Pre-bundled extensions that ship with Zed but require explicit steps to be activated and to install language servers
3. Extensions in the https://github.com/zed-industries/extensions repository that need to be explicitly activated too

Once extensions were added to Zed, the goal was to move everything in (1) to (2) or (3), but we haven't finished that work yet, since (I think, not 100% sure, will have to ask internally!) that might require some API changes for extensions or additional work, and between everything else, we decided that extensions are in a good enough spot right now. So we paused that work and went to work on other things (Linux support, for example).

That means once we finished some other projects and have more capacity again, we can come back to this and either move built-in support to extensions and/or provide more control over when to install what and how.


(My 100% personal, highly-subjective view is that it's tricky to strike a good balance here between great out of box experience, security, and not throwing pop-ups in user's faces)

**Comment by panekj** [comment]
> Instead of acting like they spat on the grave of his ancestors. It's not hard to be a little more diplomatic.

That goes for you too, instead you started your interaction with:

> Oh no, whole 14MB? And what is this NodeJS thing it's downloading? Sounds fishy...

But given your GH profile, it's not surprising.

**Comment by crabdancing** [comment]
@mrnugget

An extension system would def be nice and help compete with VSCode, so I can understand why it's a long term development goal! Is there any plan to have a text-configurable mechanism for specifying where Zed looks for the binaries? A fair few users would benefit a lot from such a mechanism -- especially for declarative distros like NixOS and Guix, and semi-reproducible packaging, e.g., flatpak.

**Comment by mtorromeo** [comment]
> I am going to put forward a hypothesis that this is a really really niche use case and 99.99% of users just want to open a JS/TS file and have it work. And the way OP phrased it ("it's COMPLETELY UNACCEPTABLE") is just nasty, entitled, and tries to make it much more of an issue than it has any right to be. Then someone posts it on reddit to stir up an unnecessary flamewar, just because.

By that reasoning, it might as well also download a full-fledged browser, otherwise opening external URLs might not just work for everyone.

I don't see why I should have yet another node installation in ~/.local/share/zed/node taking up 300M of disk space when I already have a perfectly functioning one installed on my system.

And this is not event taking into consideration the fact that it is not even working for some people, like it was reported.

At least on Linux, packages for zed can be created with all the necessary dependency requirements already defined, so there is no reason to expect that not auto-downloading node will result in a worse experience if an option is provided to disable this behavior.

Also it could at least try to detect if something is already present on the system before deciding to just download another copy of whatever dependency...

**Comment by nukeop** [comment]
> > I am going to put forward a hypothesis that this is a really really niche use case and 99.99% of users just want to open a JS/TS file and have it work. And the way OP phrased it ("it's COMPLETELY UNACCEPTABLE") is just nasty, entitled, and tries to make it much more of an issue than it has any right to be. Then someone posts it on reddit to stir up an unnecessary flamewar, just because.
> 
> By that reasoning, it might as well also download a full-fledged browser, otherwise opening external URLs might not just work for everyone.
> 
> I don't see why I should have yet another node installation in ~/.local/share/zed/node taking up 300M of disk space when I already have a perfectly functioning one installed on my system.
> 
> And this is not even taking into consideration the fact that it is not even working for some people, like it was reported.
> 
> At least on Linux, packages for zed can be created with all the necessary dependency requirements already defined, so there is no reason to expect that not auto-downloading node will result in a worse experience if an option is provided to disable this behavior.
> 
> Also it could at least try to detect if something is already present on the system before deciding to just download another copy of whatever dependency.

There's a lot of things that could be done, in principle, but the development team has to decide which of those are not worth doing because they're too niche. That's the whole point. Pressuring them to support people who are OUTRAGED that there's a whole 14MB of binaries downloaded by calling in a pitchforks-and-torches mob from reddit is just low.

**Comment by mrnugget** [comment]
> An extension system would def be nice and help compete with VSCode, so I can understand why it's a long term development goal!

We already have extensions! :) Use `zed: extensions` in the command palette.

> Is there any plan to have a text-configurable mechanism for specifying where Zed looks for the binaries? A fair few users would benefit a lot from such a mechanism -- especially for declarative distros like NixOS and Guix, and semi-reproducible packaging, e.g., flatpak.

There are no concrete plans yet.

For some language servers you can already configure where to look for the binary (Go, Rust, zig, for example) and they also look in `PATH` and won't download anything. We haven't come up with a generic API for that though since language ecosystems differ a lot in how they manage language servers (compare Rust's `rust-analyzer` binary with Ruby's way of installing language servers inside/outside `Gemfile`-using projects).

Flatpak (and with it the idea of having full control over which binaries are started where and when) came up with the port to Linux and we punted on it so far, since it's a tremendous amount of work.

But with multiple different areas of Zed — tasks, terminal, language servers, flatpak — all coming back to this one question (how do we run binaries and with which env vars and in which directory?) I think we'll need to work something out. I personally want to work on that whole environment-variable management soon (since I'm using Nix and direnv myself in my private projects)

**Comment by peterlandry** [comment]
Thanks for the response. w.r.t. this:

> (My 100% personal, highly-subjective view is that it's tricky to strike a good balance here between great out of box experience, security, and not throwing pop-ups in user's faces)

This seems pretty simply solved by a behavior setting, doesn't it? 99% of people are OK with the smooth out of the box experience, and the people that are more cautious can opt-out of the default behavior. Unless there's an aspect I'm missing?

**Comment by mtorromeo** [comment]
> Pressuring them to support people who are OUTRAGED that there's a whole 14MB of binaries downloaded by calling in a pitchforks-and-torches mob from reddit is just low.

Maybe there's a mob with pitchforks elsewhere, but that's not what I'm seeing here in this thread.

Suggesting that people are dumb for not knowing what Node.js is like, your initial remark did, is what can get people pissed.

Also, since it seems that developers did not want to change this behavior, it would be a good first step to recognize that there are legitimate concerns/issues so that a solution can be found.

The solution can be provided by the community as PRs if the developers are too busy with other priorities, but it needs to be estabilished first what they would accept as a solution.

**Comment by mrnugget** [comment]
> This seems pretty simply solved by a behavior setting, doesn't it? 99% of people are OK with the smooth out of the box experience, and the people that are more cautious can opt-out of the default behavior. Unless there's an aspect I'm missing?

No, you're right. I'm just thinking of the details: do we then also ask for an update each time? Do we ask before checking the update too (which invokes `go` tool, for example) and _then_ before downloading the update? Do we disable the language server before opening the settings for the first time and opting-in to that?

**Comment by sambonbonne** [comment]
> 99% of people are OK with the smooth out of the box experience, and the people that are more cautious can opt-out of the default behavior.

Well, I really don't like opt-out because it means I have to figure out that I need to opt-out for something I don't want. And even if I know in advance I'll need to opt-out, it forces me to go in settings and look for a specific option at first launch, which is not a great UX.

But that's just my opinion and may not reflect how every "cautious developers" think (it's not sarcasm, I think I'm really picky on these things).

**Comment by mtorromeo** [comment]
I think a confirmation dialog that says something like "Zed wants to download the following dependency: https://..." with options to allow or deny and an additional checkbox to never ask again would solve most possible use-cases.

**Comment by xpe** [comment]
Apologies for the pedantic note: a Language Server Protocol (LSP) isn't insecure, but a Language Server (LS) can be. Many of the comments above include "lsp" but they really mean "language server".

**Comment by xpe** [comment]
Just to step back for a moment. Surely we can all recognize that the "opt-in" versus "opt-out" debate is filled with landmines? So let's keep this in mind. For someone to "breeze past" the long-line of wreckage associated with this debate and proclaim an easy answer is naive at best.

Ok, now that I've said that, let's try to frame the key questions *carefully*. Very often, solving a problem well demands this. If this was an in-person synchronous process, I would say this:

> Let's not try to offer solutions (whether they "easy" or "obvious" or "work for 99% of users" or whatever) until the contours of the problem are well understood.

The above advice gets really hard in this context. Everybody has their own ideas. But if you are reading this, stop for a few minutes and try to write out the essence of the problem. Try to state the concerns of all parties even better than they have so far. Don't try to offer solutions just yet.

**Comment by Calandracas606** [comment]
> I am going to put forward a hypothesis that this is a really really niche use case and 99.99% of users just want to open a JS/TS file and have it work. And the way OP phrased it ("it's COMPLETELY UNACCEPTABLE") is just nasty, entitled, and tries to make it much more of an issue than it has any right to be.

OP appears to be someone who is attempting to package Zed for Alpine.

From a package maintainers perspective, it is completely unacceptable, a total deal-breaker for this to be the default behaviour. 

People are relying on maintainers having vetted the software in their repositories. Many users will only install software from their distribution's repositories, because they trust that the maintainers have done their due-diligence.

Ultimately, its the maintainers for distributions that need to be convinced in order for Zed to be adopted by linux users.

**Comment by peterlandry** [comment]
@sambonbonne fair enough, reasonable point. I understand there's a lot of nuance to think about, but I don't see why it could stand in the way of a behavior where on the first need to grab something, zed asks, and records the answer. It's a good start while more in-depth solutions are discussed (rather than punting on the whole issue).

**Comment by alerque** [comment]
There is quite a bit of missing the forest for the trees here, and some false dichotomies.

* Nobody is freaking about 14 MB. The issue is *when* and *how* it is downloaded and the lack of transparency. This action makes the editor completely unusable on some platforms (e.g. Nix) and introduces anti-patterns and conflicts for others (many/most distro packaging).
* There is not a binary choice between "friendly to end users" and "secure". You can have your cake and eat it too. The main thing that should be considered is less of a user facing popup or preference (where opt-in vs. opt-out is an issue, as is noise & friction), but developer facing options. If Zed offered compile time configuration this could be fixed to *everybody's* satisfaction. Distros could ship a pre-configured package with all the dependencies already provided so that the user experience is just "install and run" while also not having an app that downloads (or attempts to download) binaries behind teh users back. Meanwhile if Zed wants to ship a binary package upstream that defaults to downloading things they can.

**Comment by michealp-coder** [comment]
This was also documented [here](https://github.com/zed-industries/zed/discussions/6659#discussioncomment-8765179).

Where the VSCode theme importer is silently attempting to install a Home Brew package and execute it.

---

Very concerning, so say the least.

**Comment by mrnugget** [comment]
> This isn't just a language server concern.
> 
> 1-year ago, this security concern was also documented [here](https://github.com/zed-industries/zed/discussions/6659#discussioncomment-8765179).
> 
> Where the VSCode theme importer is silently attempting to install a Home Brew package and execute it.

I wasn't around a year ago, but as far as I can tell: that extension runs in VS Code and wasn't published nor written by the Zed team. The theme importer it uses under the hood _was_ written by the Zed team and everybody can read through its source here: https://github.com/zed-industries/zed/tree/main/crates/theme_importer That `theme_importer` was _not_ published to homebrew by Zed either. (Also, @michealp-coder, you were in that thread in which it was explained how that `theme_importer` was created: https://github.com/zed-industries/zed/issues/7111#issuecomment-1918289658)

Since Zed doesn't advertise nor officially recommend that VS Code extension, I don't think there's a lot we can do about it (should we actively advertise _against_ it? not sure)

**Comment by jarcane** [comment]
> EDIT: Now I found that it downloads ([here](https://github.com/zed-industries/zed/blob/de8ef081436317689333b29324b3a64343745914/crates/supermaven_api/src/supermaven_api.rs#L189)) even some proprietary binary from https://supermaven.com/, i.e. unaudited and unauditable code, without any verification (except TLS)! At least this is not downloaded by default… I hope…

Am I the only one who noticed that Supermaven isn't just a proprietary binary, but one from an AI startup?

That is such a massive exposure surface all of its own. Some employers even have specific policies against using AI with private code these days, for just that reason.

**Comment by mrnugget** [comment]
Supermaven requires an explicit setting to be downloaded and enabled. It's not enabled by default.

**Comment by Sinofine** [comment]
Zed is not noticeably faster than VSCode, yet it has many flaws. Although it is implemented in Rust, it still relies on Node.js (It even downloads that from the internet without users permission).

**Comment by tv42** [comment]
For anyone else like me who wants to put in a technical stopgap kludge to prevent Zed from downloading things that pretty much aren't going to work on my NixOS anyway, put this in your settings.json:

```
  "server_url": "https://disable-zed-downloads.invalid",
```

**Comment by revskill10** [comment]
Allow an environment variable like `NODE_PATH` should skip the unessessary Node download.

**Comment by kurucu** [comment]
> Then here's how the OP could have looked:
> 
> > Hi, I am a security conscious developer who often maintains critical applications for major corporations. Could you please include an option to confirm downloads of the following binaries? [list of binaries] I would like to be able to download them on an opt-in basis.
> 
> Instead of acting like they spat on the grave of his ancestors. It's not hard to be a little more diplomatic.

So your argument isn't that they were wrong, just impolite? You still agree with the outcome then?

**Comment by nukeop** [comment]
> > Then here's how the OP could have looked:
> > > Hi, I am a security conscious developer who often maintains critical applications for major corporations. Could you please include an option to confirm downloads of the following binaries? [list of binaries] I would like to be able to download them on an opt-in basis.
> > 
> > 
> > Instead of acting like they spat on the grave of his ancestors. It's not hard to be a little more diplomatic.
> 
> So your argument isn't that they were wrong, just impolite? If that means you still agree with the outcome, what are you looking for?

A request isn't something you can be "wrong" or "right" about. I think it's an extremely niche thing that a small minority of users will ever care about, and something that disproportionately wastes the resources of the development team. This is why you don't want to sell to developers, especially Linux developers - they struggle to fork over the money and act like everyone owes them the Earth and the Moon.

**Comment by SichangHe** [comment]
@mrnugget, I highly suggest you hide some of the comments in this thread. Many comments are just people accusing each other instead of contributing to the discussion of how to improve the situation.

**Comment by xpe** [comment]
> For anyone else like me who wants to put in a technical stopgap kludge to prevent Zed from downloading things that pretty much aren't going to work on my NixOS anyway, putting this in your settings.json breaks at least some of that:
> 
> ```
>   "server_url": "https://disable-zed-downloads.invalid",
> ```

Until someone buys that domain and pwnage occurs.

**Comment by GoldsteinE** [comment]
> Until someone buys that domain and pwnage occurs.

You can’t buy an `.invalid` domain, it’s reserved.

**Comment by xpe** [comment]
> > Until someone buys that domain and pwnage occurs.
> 
> You can’t buy an `.invalid` domain, it’s reserved.

Ah, very nifty: https://en.wikipedia.org/wiki/.invalid

> The reasons for reservation of these top-level domain names is to reduce the likelihood of conflict and confusion. This allows the use of these names for either documentation purposes or in local testing scenarios.

**Comment by BobobUnicorn** [comment]
Has the Zed team verified that this approach is compatible with their version of the GPL? My understanding of it is that it requires any linked (dynamically, or statically) non-system libraries to also have source available, as a plugin or extension that's automatically downloaded seems to count as part of a single combined program in certain cases [ref](https://www.gnu.org/licenses/gpl-faq.html#GPLPlugins). In the case of Supermaven, I don't see any way of accessing its corresponding source.

**Comment by nukeop** [comment]
> Has the Zed team verified that this approach is compatible with their version of the GPL? My understanding of it is that it requires any linked (dynamically, or statically) non-system libraries to also have source available, as a plugin or extension that's automatically downloaded seems to count as part of a single combined program in certain cases [ref](https://www.gnu.org/licenses/gpl-faq.html#GPLPlugins). In the case of Supermaven, I don't see any way of accessing its corresponding source.

GPL places limitations on what others (licensees) can do with Zed's source. It doesn't place limitations on what the authors themselves can do with their own program.

**Comment by BobobUnicorn** [comment]
> > Has the Zed team verified that this approach is compatible with their version of the GPL? My understanding of it is that it requires any linked (dynamically, or statically) non-system libraries to also have source available, as a plugin or extension that's automatically downloaded seems to count as part of a single combined program in certain cases [ref](https://www.gnu.org/licenses/gpl-faq.html#GPLPlugins). In the case of Supermaven, I don't see any way of accessing its corresponding source.
> 
> GPL places limitations on what others (licensees) can do with Zed's source. It doesn't place limitations on what the authors themselves can do with their own program.

What this effectively means is that forking Zed would require fully analyzing what parts would have to be removed, since it can't just be redistributed in its current form from my reading of it.

**Comment by mikayla-maki** [comment]
Since 28 separate people get a ping on every message here, I'm going to lock this issue for now. We are currently working on this, follow along here: https://github.com/zed-industries/zed/pull/14034

**Comment by mikayla-maki** [comment]
Status update: We are still working on this! The major blocker is that extensions have not been setup to interact with setting. However, we also need to change this API to support our upcoming remote development feature. So we're going to roll both of these breaking changes into a larger extension update, coming this November or December :)

**Comment by mikayla-maki** [comment]
Hello everyone! As you may have noticed, I did not get a chance to close this issue in November or December. My apologies for not managing time for this correctly. However, this issue is still important to us and solving it is essential for deploying Zed in many enterprise environments. We're not going to miss this, scheduling is just messy :)

**Comment by SomeoneToIgnore** [comment]
Looks like I was wrong and https://github.com/zed-industries/zed/pull/44887 will be pivoted in a slightly lighter initial approach, with certain `node` and other tools spawned despite whatever restrictions will be implied by the future functionality.

Hence, we need to work on this issue more later.

**Comment by yonipeleg33** [comment]
How about trying to detect the existence of the required binaries in the user's environment? Was this approach discussed somewhere?

The assumption behind such an approach can be "Zed users have the required binaries for their workflows, so Zed doesn't need to ship its own".  And if Zed does need a binary it didn't find, it can prompt the user with a one-click approval popup to download it, being transparent about the source and purpose of this binary.

If this approach sounds interesting, the technicalities can be further discussed

**Comment by imbev** [comment]
> How about trying to detect the existence of the required binaries in the user's environment? Was this approach discussed somewhere?

This would be especially useful for Nix users. On NixOS (and older distributions), many extensions may fail because they download binaries expecting unavailable shared libraries.

**Comment by darix** [comment]
or who says that the binaries that are automatically downloaded by zed arent compromised. supply chain attacks are so prevelant lately. Users should have the control to point zed to existing binaries or at least being asked for downloads. or have an option to disable them. and nodejs runtime is just the top of the iceberg. 

I currently block zed from downloading anything by blocking write access to the needed paths via apparmor. It is also a bit annoying that zed can not use the distro packaged treesitter libraries and has to ship its own copies again. same goes for LSP servers.

**Comment by yonipeleg33** [comment]
Seeing how my [suggestion](https://github.com/zed-industries/zed/issues/12589#issuecomment-4524892558) is getting positive feedback overall, I'd like to propose a practical way forward, focusing on scoped, low-risk incremental changes:

_Note: I'm calling NodeJS binary, LSPs, npm packages, etc. "tools" here_

---

**Edit:**
I looked into the code a bit now, and seems like a decent amount of "discovery" is already in place; so I'd say step 1 is mainly to instrument the already-existing discovery points, then re-evaluate the situation from there.

---

### Step 1: Metrics only

Add a reporting layer to the code that's downloading the tools (haven't looked that up in the codebase yet so I don't have specific links, but I'd imagine it's relatively scoped and easy to find).

The reporting layer will run the tools discovery exactly as it intended to be used, but only report back basic metrics, such as:
- success/failure
   - if failed, report the reason (e.g. permission denied, not found, etc.)
- version found
- tool path

These metrics fit nicely under the [client-side-metrics](https://github.com/zed-industries/zed/blob/main/docs/src/telemetry.md#client-side-metrics) IMO.

**Criteria for moving forward:** success rate above some threshold (I'd go with 90-95%, but TBD)
Iterate on the tool discovery logic until that threshold is hit.

### Step 2: Silent fallback to existing logic

Try to use the discovered tool, but if any error occurs (either during the discovery phase or during actual runtime usage), silently fall back to downloading from Zed's existing sources.

While this step does not achieve the end goal yet, it also doesn't introduce any user-facing regression, as downloading "silently" is the current behavior.

This step should also report more or less the same basic metrics as in step 1.

**Criteria for moving forward:**  success rate above some threshold (now success/failure represent actual usage, not just hypothetical discovery)

### Step 3: Approval pop-up

Instead of silently falling back to downloading, prompt the user with a pop-up asking to download the tool.
The pop-up should be clear about the source (download url), the destination, and the purpose of this tool (i.e. what Zed uses this tool for).

### Step 4: User-provided tool paths

Allow users to configure the path, and maybe the full command Zed should run for using each tool.
Something like:
```json
{
  "tools": {
    "node": {
      "command": "/path/to/node",
      "arguments": []
    }
  }
}
```

---

This is just one possible way forward. I’m trying to be a bit more proactive and help unblock this two-year-old issue. Of course, I’m open to feedback and discussion, and I’d appreciate your thoughts.

**Comment by ZtereoHYPE** [comment]
While I see the niceness of the "it just works" factor, I really don't think Zed should download all the language servers for the languages mentioned in the issue by default. Perhaps when launching for the first time, or in settings, there could be a simple list that mentions what binary should be downloaded and from where, allowing the user to supply its own via a path? 

I envision a GUI popup with checkboxes for the desired features, and a large button to proceed with the download. That way, the nice ootb experience can be almost fully maintained, as the user can simply press the button to proceed as usual.

While the tool discovery approach mentioned above might work, it's not transparent to the developers, and would result in lots of additional logic which can simply be avoided (less is more :p )

**Comment by yonipeleg33** [comment]
@ZtereoHYPE I'm assuming disabling or downgrading the "default just works" experience is out of the discussion, but I could be wrong of course.
If I'm right though, the most I can think of to take your direction is to prompt the user with the default sources for everything (like my "step 3")

> While the tool discovery approach mentioned above might work, it's not transparent to the developers, and a entails lot of additional logic which can simply be avoided (less is more :p )

1. I looked through the code a bit, and there is a decent amount of tool discovery work already; node and LSP binaries are attempted to load from the user's PATH before downloading.
   (It could be the case that the current code isn't working for all setups, which might explain why this issue gets a decent amount of traction; in any case, instrumenting the existing code is the first step IMO).
2. Discovering existing binaries in the user's setup (PATH or otherwise) is a common practice that developers can expect in general.  I don't think there's a transparecy issue here.

**Comment by darix** [comment]
zed works already very well as an editor without nodejs and prettier and the other things it tries to download without asking.

**Comment by robindegen** [comment]
I'm also very concerned about this. I'd rather have something not working and make the concious decision to leave it that way; than getting hit by the next inevitable supply chain attack. I only use rust; i dont need anything else. And if that means i still need npm, then i'd rather code without a language server at all.

**Comment by yonipeleg33** [comment]
Valid concerns indeed.
Ultimately, that’s a decision for the team, so I’ll leave it to the maintainers to take a stance.

**Comment by andreashaerter** [comment]
This is really important. There are many users which are fed up with VSCode because of

1. [The politics](https://ghuntley.com/fracture/) and
2. the extension security ([the recent events](https://github.com/microsoft/vscode/issues/52116) again shed light on this).

I really want to switch to Zed because of the speed, the UX and the developer team. And I do not want to support VSCode anymore.

I know a lot in my peer-group feeling the same. But to push people giving up an working ecosystem without at least improve the security while giving up a lot of integration won't work if things like this are still unresolved issues. :(

**Comment by flukejones** [comment]
One of the reasons I switched to Zed was the appeal of a lightweight editor, both in size and process - it mostly achieves that except for the downloading of Node separate to whatever the running OS may have and the resulting 1.5GB of cache from that.

Personally I don't do any work in environments that require LSP that run on node and wish I could fully disable Node plus everything that relies on it (even if it breaks some plugins)

It occurs to me that there really should be an option to block all external downloads such as those from extensions - but this would require a bit of a change in how some plugins are written.. They would all need to ensure that a path can be set for their required bins/files.

**Comment by alerque** [comment]
Zed's long term resistance to addressing this and other similar issues is why the [Gram](https://gram-editor.com/) editor was forked. There is also the [Zedless](https://zedless.org/) project which has some overlap in goals but takes a different approach to getting there. If you're into Zed's interface and features but don't like the AI clutter and unsupervised network actions, I can suggest giving Gram a shot.

**Comment by robindegen** [comment]
>   Zed's long term resistance to addressing this and other similar issues is why the [Gram](https://gram-editor.com/) editor was forked. There is also the [Zedless](https://zedless.org/) project which has some overlap in goals but takes a different approach to getting there. If you're into Zed's interface and features but don't like the AI clutter and unsupervised network actions, I can suggest giving Gram a shot.

thanks ill try those

**Comment by javierguerragiraldez** [comment]
this is the second time I've tried Zed.  the first time, i killed it when I noticed it was downloading node.js  (which is almost at the top of technologies I don't want in my computer)  Now, I was careful to install it without any "optional" dependencies, and for a few hours, it seemed to behave very well; until I opened a Python project, and noticed it activated a node-based LSP.  I checked again and, yes, it had downloaded it silently.  I removed it and it was downloaded again.  no questions, no alerts, not even a spinning wheel.

sorry, but I can't accept that kind of contempt to my choices.

If Zed sometime becomes usable without pulling arbitrary code, I might try again.  But I'm not holding my breath.

**Comment by jangorecki** [comment]
Don't need to wait for zed. I tried zed's fork, gram editor (mentioned 3 posts above) and it works great.
If matter like not-auto-downloading node.js is such a low priority for zed then we should not expect much from this project.
I wish all the best to zed project but considering they are pushing features for users that we don't want, violating good security practices, then obviously I prefer to stay on a successful fork of zed.

**Comment by nikaro** [comment]
~~@JosephTLyons this is issue is absent from the [Top-Ranking Issues](https://github.com/zed-industries/zed/issues/5393) despiste its >=153 upvotes (it should be at the 19th place). There probably is a bug with the ranking script.~~

Edit: 🤦 i looked at the "Features" not the "Bugs"...

**Comment by holly-hacker** [comment]
> [@JosephTLyons](https://github.com/JosephTLyons) this is issue is absent from the [Top-Ranking Issues](https://github.com/zed-industries/zed/issues/5393) despiste its >=153 upvotes (it should be at the 19th place). There probably is a bug with the ranking script.

It is marked second under Bugs, which matches this issue's Bug tag which the OP added.

**Comment by noctisatrae** [comment]
Honestly, I kinda lost my mind when I realized that Zed just installed a random `rust-analyzer` binary without telling me. I wish I could just say no and have time to properly install/configure it through `rustup`. My computer does not have that much storage, and I hate having to chase down and clean cache folder. 

Can you guys make an option to forbid downloading things from the internet? I'd rather have it not work and have to install it myself than an incorrect silent install.

**Comment by lnicola** [comment]
That binary is the recommended way to install rust-analyzer, not rustup.

**Comment by alerque** [comment]
It doesn't matter what the recommended way to install rust-analyzer is. Having a random editor download and install it's own copy behind the user's back is egregiously bad behavior even if it downloads the upstream recommended version of the binary.

**Comment by teohhanhui** [comment]
> That binary is the recommended way to install rust-analyzer, not rustup.

Nowhere does it actually say that:

https://rust-analyzer.github.io/book/rust_analyzer_binary.html

`rustup` is even there on the page itself as one of the officially supported installation methods.

**Comment by alerque** [comment]
@nukeop There are literally **hundreds** of language server binaries out there: half a dozen at least for Python and a few more for C and no telling how many for a long tail of other languages. All the editor needs to do is say when you open a file in a given language is "hey if you install X I can enable LSP support". Silently downloading and installing binaries for every/any language file I open is egregious. Next you're going to tell be it should also install `cargo` and `rustc` and `clang` and `llvm` and `mold` and who knows what else because those are common bits of toolchain and called directly by functions in the editor. Just no.

**Comment by dotprod** [comment]
Ideas to consider:
- obligatory user consent with full risk explanation;
- editor setting that specifies minimal possible time since last package update, to prevent chain-of-supply attacks, (for me its 1 week at least);
- mandatory pin of any package versions, update can download only them;
- mandatory CVE monitoring and reporting;

**Comment by thomas-mckay** [comment]
My preferred option would be to have a way to disable downloads entirely, and a way to configure binary paths manually, so I can hook zed into my own setup. If those configs had path translation (host to container), or if it supported containers natively, so I could run tools (LSP...) containerized, that'd be a big plus.

**Comment by radupotop** [comment]
There is a way to configure LSP paths, so that Zed uses system-installed LSPs. If you specify the following in your settings file:

```
"lsp": {
"pyright": {
  "binary": {
    "path": "/usr/bin/pyright-langserver",
    "arguments": ["--stdio"],
  },
},
...
```

Then Zed will use that system package instead of fetching it over the internet. It works for all LSPs. 
But more recently I've found that Zed picks up system-installed LSPs directly without any need for configuration.

The same works for agent_servers:

```
"agent_servers": {
  "codex-acp": {
    "type": "custom",
    "command": "/usr/bin/codex-acp",
    "default_config_options": {
    ...
    },
  },
},
```

**Comment by robindegen** [comment]
> There is a way to configure LSP paths, so that Zed uses system-installed LSPs. If you specify the following in your settings file:
> 
> ```
> "lsp": {
> "pyright": {
>   "binary": {
>     "path": "/usr/bin/pyright-langserver",
>     "arguments": ["--stdio"],
>   },
> },
> ...
> ```
> 
> Then Zed will use that system package instead of fetching it over the internet. It works for all LSPs. 
> But more recently I've found that Zed picks up system-installed LSPs directly without any need for configuration.
> 
> The same works for agent_servers:
> 
> ```
> "agent_servers": {
>   "codex-acp": {
>     "type": "custom",
>     "command": "/usr/bin/codex-acp",
>     "default_config_options": {
>     ...
>     },
>   },
> },
> ```

that's a good start. but typically this means its too late, since typically you'd install the plugin first or simply just launch it, then it downloaded potentially trouble already. switching to system lsp afterwards is too late. im also not sure what the exact config is for each of the plugins or how to even find out

**Comment by tesujimath** [comment]
It shouldn’t be necessary, and indeed it is an anti-pattern, to specify the full path to the desired tools, because `$PATH`. (And it precludes use of direnv for tool installation.)

The Zed developers should ask themselves whether they are targeting beginners, for whom auto-installation is a huge convenience, or advanced users, for whom it is anathema.

**Comment by stolen-byte** [comment]
Given the "security & privacy" tag it's a little worrying that this issue has had so little movement in over 2 years, and all PR's closed.

A little drastic, but i would have thought the immediate easy option would have been to just remove all downloads entirely and require manual installation, no if's or but's, the new workspace-trust thing just doesn't cut it, it is way too all-or-nothing, yes it (maybe?) stops the downloads, but i also cannot run any manually installed LSP's, and no option to allow just those...

Given node/npm's tendency to be the first port of call for supply chain attacks, i myself have actually gone so far as creating my own little "fake" node program, which gets installed as `node`/`npm` binaries, answers requests for `--version` (to satisfy any "is it installed" queries), but otherwise logs the caller process + arguments to syslog then exits with an error. (Side note: i use Deno for everything now, and doing this was the only way i could have zed installed without nodejs installed).

Granted, that doesn't deal with the other stuff, like supermaven mentioned by the OP, but i've had no issues thus-far, i don't get any LSP for JSON files...big deal, who even needs that? Small price to pay for not having a secret stash of 1GB+ npm packages hidden in the zed data directories.

Also really annoying that it doesn't pick up `prettier` from my `$PATH` either (it can be installed via deno).

**Comment by cedws** [comment]
Just wanted to share my hack, I created a [Tetragon](https://tetragon.io/) policy that SIGKILLs anything that Zed tries to execute in `.local/share/zed/languages` and `.local/share/zed/remote_extensions`. I use Nix to install the language servers + Node globally. I think this is superior to trying to block downloads with `chattr` or something because Zed will presumably waste bandwidth trying to keep downloading.

I'm doing this out of concern for supply chain security. These days it's absolutely crucial everything is pinned, and as far as I can see Zed extensions don't pin binaries they download from GitHub. This is very risky in today's landscape where repos are being compromised all the time and there's practically zero uptake of Immutable Releases.

Also, using Nix to manage these tools instead gives me reproducibility.

I was hopeful that I'd be able to prevent Zed doing npm installs/binary downloads using the new extension capabilities, but for some reason this doesn't seem to apply when using remote workspaces.

**Comment by darix** [comment]
> I created a [Tetragon](https://tetragon.io/) policy that SIGKILLs anything that Zed tries to execute in `.local/share/zed/languages` and `.local/share/zed/remote_extensions`.

my apparmor profile just denies writing to all the places zed would download things too. it is a bit whine-y about it on start up. and basically the only thing it can launch is git. and the git launched by zed can only write to certain places. :) 

Zed is often not happy with me.

**Comment by spotlesscoder** [comment]
It's completely unacceptable for me that zed installs such risky supply-chain attack vulnerable software without asking for my consent first!

**Comment by 0x08088405** [comment]
This behavior has actively deterred me from wanting to touch Zed at all, and the fact this has not been addressed in years is concerning for the security standards they're trying to deliver, which almost every other text editor meets by doing absolutely nothing of the sort. A "solution" is running Zed on NixOS without nix-ld, such that no binary it downloads will be able to execute, as it's downloading for example a dynamically linked clangd, but it does not stop it from downloading 220MiB (!!!) without asking at all, which is especially rude on metered connections. Please do not download and especially run binaries without my consent. Hoping this is addressed sometime. If, for some reason, a non-negotiable goal for a programming text editor is to cater to people with zero technical skills, who really cannot download a binary and specify a path or put it in their `$PATH`, at least add a prompt, and a way for me to specify "never".

**Comment by reivilibre** [comment]
> running Zed on NixOS without nix-ld

I advise against this, as Zed seems to re-download it every relaunch (iirc) because it detects that it's broken. That's a particularly nice way to use up your limited data plan.

**Comment by Maxi9090** [comment]
This issue forces me to use this workaround: <https://github.com/Maxi9090/zed-customlsp>

<details>
<summary>And with these settings as an example</summary>

```json
{
  "enable_language_server": false,
  "formatter": "none",
  "auto_install_extensions": {
    "html": false
  },
  "lsp": {
    "superhtml": {
      "binary": {
        "path": "C:/tools/superhtml.exe",
        "arguments": ["lsp"]
      }
    }
  },
  "languages": {
    "HTML": {
      "enable_language_server": true,
      "language_servers": ["superhtml"],
      "formatter": "language_server"
    }
  }
}
```
</details>

It wasn't easy finding this proper workaround, but I got desperate enough because I got so annoyed that most LSP extensions are slow to start up, constantly auto-updates, and refuses to run without successfully checking for no new updates.

After setting this all up with the LSP servers I need, Zed got a lot less network traffic, and a lot less surprises in my Zed user data folder. I don't miss the LSP servers by the extensions.

But that's not enough. I want to control what Zed downloads without these workarounds, or at least the possibility to do that. I don't want all these extras, automatically downloaded stuff, node and all that before I can comprehend what Zed is doing on a new install with no config. I don't want to have to manually remove these unwanted files with a file manager deep in some odd place every time, and I don't want to have to use some special security software to secure this attack vector.

I use Zed because I like it better than other IDEs, but this issue is one of the disadvantages with Zed at the moment.

**Comment by stolen-byte** [comment]
/sigh

It seems in this age of "we must have all the AI", and "just let the AI do it", we're just shouting into the void, as it were.
I guess it's back to basic text-editors for the forseeable future, because lets face it, without LSP's (ie: worktree trust enabled) that's all it basically is, a fancy-looking notepad. 🫨

**Comment by andreashaerter** [comment]
> It seems in this age of "we must have all the AI", and "just let the AI do it", we're just shouting into the void, as it were.

Simply off-topic.

There are enough AI usage and features issues to discuss. This one is completely unrelated, and (if you look at the labels) not unrecognized.

Don't get me wrong: This one is the only blocker for my org to completely switch over to Zed (like I personally) did. But it does not help anybody to rant about AI in unrelated issues so everybody subscribed get's email notifications.

**Plus: [`{ "disable_ai": true }` exists.](https://zed.dev/blog/disable-ai-features)** in Zed

**Comment by robindegen** [comment]
> There are enough AI usage and features issues to discuss. This one is completely unrelated, and (if you look at the labels) not unrecognized.

They can put all the labels they want but it's been open for 2 years now. That label means nothing apparently.

**Comment by andreashaerter** [comment]
> They can put all the labels they want but it's been open for 2 years now. That label means nothing apparently.

Well, at least it is not closed :) and some tags, especially severity, were added in December 2025 plus it was moved to https://github.com/orgs/zed-industries/projects/78 in March 2026. So I don't think it means nothing, just slow progress = we have to wait or one has to work himself on this issue (in the end, it's OSS).

Still, don't get me wrong: I would also be *really* happy if this issue and https://github.com/zed-industries/zed/discussions/24737 would be solved; this one here is our adoption blocker and the other one would justify sending real company money to Zed. I just wanted to state that AI rants won't help to solve #12589.

**Comment by stolen-byte** [comment]
> it does not help anybody to rant about AI in unrelated issues so everybody subscribed get's email notifications.

I would hardly call 2 sentences a "rant" 😆 

Anyway, that was meant more as a dig at the fact there's been no official comment on this for over 2 years, and most bugfixes pertaining to agent stuff are fixed almost immediately.

And then there's frustration, needed to fix a bug in an old game mod (C# project), decided to try out the roslyn language-server to see if it was working at all yet, so proceeded to `dotnet tool install -g` it, then upon opening zed, was greeted by:

<img width="334" height="40" alt="Image" src="https://github.com/user-attachments/assets/fdaaaf70-f53b-410e-be93-279dfae79f2d" />

Needing me to
- manually override the binary in settings, and
- figure out what command-line flags i was supposed to pass to it.
- eventually figure out it still doesn't even work (the LSP), so it was all a waste anyway :(

**Comment by matanelevi** [comment]
WOW.
Thank you for sharing.

This is unbelievable.. major security issue.
I will remove Zed, unfortunately.
